tests/test_server.py

```diff
@@ -239,6 +239,28 @@ def test_query_promotion_list_by_name(self):
         data = json.loads(resp.data)
         query_item = data[0]
         self.assertEqual(query_item['product_id'], 9527)
+		
+    def test_query_promotion_list_by_product_id(self):
+        """ Query Promotions by Product id """
+        resp = self.app.get('/promotions', query_string='product_id=9527')
+        self.assertEqual(resp.status_code, status.HTTP_200_OK)
+        self.assertGreater(len(resp.data), 0)
+        self.assertIn('9527', resp.data)
+        self.assertNotIn('9526', resp.data)
+        data = json.loads(resp.data)
+        query_item = data[0]
+        self.assertEqual(query_item['name'], '20%OFF')
+		
+    def test_query_promotion_list_by_discount_ratio(self):
+        """ Query Promotions by Discount ratio """
+        resp = self.app.get('/promotions', query_string='discount_ratio=0.8')
+        self.assertEqual(resp.status_code, status.HTTP_200_OK)
+        self.assertGreater(len(resp.data), 0)
+        self.assertIn('0.8', resp.data)
+        self.assertNotIn('0.2', resp.data)
+        data = json.loads(resp.data)
+        query_item = data[0]
+        self.assertEqual(query_item['product_id'], 9527)
```

Comment:
Does there test cover `by_promotion_id` and the `else` branch?

Reply:
These tests cover all the else branch, and, I don't think a query by user can get the promotion_id because even us won't use it to create a new promotion, so I delete it.